pkg/secrets-store/utils.go

```diff
@@ -0,0 +1,790 @@
+// If a secret with the same name already exists in the namespace of the pod, it's updated.
+func createOrUpdateK8sSecret(ctx context.Context, name string, namespace string, datamap map[string][]byte, secretType corev1.SecretType) error {
+	// recreating client here to prevent reading from cache
+	c, err := getClient()
+	if err != nil {
+		return err
+	}
+	secretKey := types.NamespacedName{
+		Namespace: namespace,
+		Name:      name,
+	}
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+		},
+		Type: secretType,
+	}
+
+	err = c.Get(ctx, secretKey, secret)
+	if err != nil {
+		log.Error(err, "error from c.Get for secret: %s, ns: %s", name, namespace)
+		if errors.IsNotFound(err) {
+			secret.Data = datamap
+			if err := c.Create(ctx, secret); err != nil {
+				log.Error(err, "error while creating K8s secret: %s, ns: %s", name, namespace)
+				return err
+			}
+		} else {
+			log.Error(err, "error while retrieving K8s secret: %s, ns: %s", name, namespace)
+			return err
+		}
+	}
+	secret.Data = datamap
+	if err := c.Update(ctx, secret); err != nil {
+		log.Error(err, "error while updating K8s secret: %s, ns: %s", name, namespace)
+		return err
+	}
+
+	log.Infof("created k8s secret: %s, ns: %s", name, namespace)
+	return nil
+}
+
+// deleteK8sSecret deletes a secret by name
+func deleteK8sSecret(ctx context.Context, name string, namespace string) error {
+	// recreating client here to prevent reading from cache
+	c, err := getClient()
+	if err != nil {
+		return err
+	}
+	secretKey := types.NamespacedName{
+		Namespace: namespace,
+		Name:      name,
+	}
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+		},
+	}
+
+	if err := c.Get(ctx, secretKey, secret); err != nil {
+		if errors.IsNotFound(err) {
+			log.Infof("k8s secret not found during delete. Skip. secret: %s, ns: %s", name, namespace)
+			return nil
+		}
+		log.Error(err, "error while getting K8s secret: %s, ns: %s", name, namespace)
+		return err
+	}
+
+	if err := c.Delete(ctx, secret); err != nil {
+		log.Error(err, "error while deleting K8s secret: %s, ns: %s", name, namespace)
+		return err
+	}
+
+	log.Infof("deleted k8s secret: %s, ns: %s", name, namespace)
+	return nil
+}
+
+// setStatus adds pod-specific info to byPod status of the secretproviderclass object
+func setStatus(ctx context.Context, obj *unstructured.Unstructured, id string, namespace string) error {
+	log.Infof("setStatus for pod: %s, ns: %s", id, namespace)
+	// recreating client here to prevent reading from cache
+	c, err := getClient()
+	if err != nil {
+		return err
+	}
+	status := map[string]interface{}{
+		"id":        id,
+		"namespace": namespace,
+	}
+	statuses, exists, err := unstructured.NestedSlice(obj.Object, "status", "byPod")
+	if err != nil {
+		return err
+	}
+	if !exists {
+		if err := unstructured.SetNestedSlice(
```

Comment:
We are setting the nested slice here - https://github.com/kubernetes-sigs/secrets-store-csi-driver/pull/186/files#diff-a7cc474b343f3d0bb08dd248f288bacdR466-R470. Can we remove this in that case?